docs/proposals/202004_embedd_cortex_frontend.md

```diff
@@ -0,0 +1,121 @@
+* Reuse as much as possible between projects, contribute.
+
+## No Goals
+
+* Create Thanos specific response caching from scratch.
+
+## Proposal
+
+The idea is to create `thanos frontend` component that allows specifying following options:
+
+* `--query-range.split-interval`, `time.Duration`
+* `--query-range.max-retries-per-request`, `int`, default = `5`
+* `--query-range.disable-step-align`, `bool`
+* `--query-range.response-cache-ttl` `time.Duration` default = `1m`
+* `--query-range.response-cache-config(-file)` `pathorcontent` + [CacheConfig](https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/store/cache/factory.go#L32)
+
+We plan to have in-mem, fifo and memcached support for now. Cache config will be exactly the same as the one used for Store Gateway.
+
+### Open Questions
+
+* Is `thanos frontend` a valid name? I feel it's short and verbose enough.
+
+### Alternatives
+
+#### Don't add anything, document Cortex query frontend and add examples of usage
+
+Unfortunately we tried this path already without success. Reasons were mentioned in [Motivation](202004_embedd_cortex_frontend.md#Motivation)
+
+#### Add response caching to Querier itself, in the same binary.
+
+This will definitely simplify deployment if Querier would allow caching directly. However, this way is not really scalable.
+
+Furthermore, eventually frontend will be responsible for more than just caching. It is meant to do query planning like splitting or even
+advanced query paralization (query sharding). This might mean future improvements in terms of query scheduling, queuing and retrying.
+This means that at some point we would need an ability to scale query part and caching/query planner totally separately.
+
+Last but not least splitting queries allows to perform request in parallel. Only if used in single binary we can achieve load balancing of those requests.
+
+NOTE: We can still consider just simple response caching inside Querier if user will request so.
+
+#### Write response caching from scratch.
+
+I think this does not need to be explained. Response caching has proven to be not trivial. It's really amazing that we
+have opportunity to work towards something that works with experts in the field like @tomwilkie and others from Loki and Cortex Team.
+
+Overall, [Reusing is caring](https://www.bwplotka.dev/2020/how-to-became-oss-maintainer/#5-want-more-help-give-back-help-others).
+
+## Work Plan
+
+1. Refactor [IndexCacheConfig](https://github.com/thanos-io/thanos/blob/55cb8ca38b3539381dc6a781e637df15c694e50a/pkg/store/cache/factory.go#L32) to generic cache config so we can reuse.
+1. Add necessary changes to Cortex frontend
+    * Metric generalization (they are globals now).
+1. Add `thanos frontend` subcommand.
+1. Add proper e2e test using cache.
+1. Document new subcommand
+1. Add to [kube-thanos](https://github.com/thanos-io/kube-thanos0)
```

Comment:
Minor typo at the en of URL here. Should be `/kube-thanos` instead `/kube-thanos0` :)